src/smc-webapp/course/course_funcs.coffee

```diff
@@ -0,0 +1,67 @@
+# Pure functions used in the course manager
+
+# SMC libraries
+misc = require('smc-util/misc')
+{defaults, required} = misc
+
+immutable = require('immutable')
+
+# Transforms Iterable<K, M<i, m>> to [M<i + primary_key, m + K>] where primary_key maps to K
+# Dunno if either of these is readable...
+# Turns Map(Keys -> Objects{...}) into [Objects{primary_key : Key, ...}]
```

Comment:
I don't even know if this notation makes sense. Either one also reads kind of like code.